packages/lending/src/protocols/morphoblue/lending-protocol.ts

```diff
@@ -0,0 +1,229 @@
+
+  get oracleIface() {
+    if (!this._oracleIface) {
+      this._oracleIface = Oracle__factory.createInterface();
+    }
+    return this._oracleIface;
+  }
+
+  private _irmIface?: IrmInterface;
+
+  get irmIface() {
+    if (!this._irmIface) {
+      this._irmIface = Irm__factory.createInterface();
+    }
+    return this._irmIface;
+  }
+
+  getMarketName(id: string) {
+    const { loanToken, collateralToken } = getMarket(this.chainId, id);
+    return `${DISPLAY_NAME} ${loanToken.symbol}(${collateralToken.symbol} collateral)`;
+  }
+
+  async getPortfolio(account: string, marketId: string) {
+    const { loanToken, collateralToken, loanTokenPriceFeedAddress, oracle, irm, lltv } = getMarket(
+      this.chainId,
+      marketId
+    );
+
+    const calls: common.Multicall3.CallStruct[] = [
+      {
+        target: this.morpho.address,
+        callData: this.morphoIface.encodeFunctionData('position', [marketId, account]),
+      },
+      {
+        target: this.morpho.address,
+        callData: this.morphoIface.encodeFunctionData('market', [marketId]),
+      },
+      {
+        target: loanTokenPriceFeedAddress,
+        callData: this.priceFeedIface.encodeFunctionData('latestAnswer'),
+      },
+      {
+        target: oracle,
+        callData: this.oracleIface.encodeFunctionData('price'),
+      },
+    ];
+
+    const { returnData } = await this.multicall3.callStatic.aggregate(calls, { blockTag: this.blockTag });
+
+    const { borrowShares, collateral } = this.morphoIface.decodeFunctionResult('position', returnData[0]);
+
+    const { totalSupplyAssets, totalSupplyShares, totalBorrowAssets, totalBorrowShares, lastUpdate, fee } =
+      this.morphoIface.decodeFunctionResult('market', returnData[1]);
+
+    const [_loanTokenPrice] = this.priceFeedIface.decodeFunctionResult('latestAnswer', returnData[2]);
+
+    // It corresponds to the price of 10**(collateral token decimals) assets of
+    // collateral token quoted in 10**(loan token decimals) assets of loan token
+    // with `36 + loan token decimals - collateral token decimals` decimals of precision.
+    const [normalizedCollateralLoanPrice] = this.oracleIface.decodeFunctionResult('price', returnData[3]);
+    const normalizedCollateralLoanDecimals = 36 + loanToken.decimals - collateralToken.decimals;
+
+    const loan = totalBorrowAssets.mul(borrowShares).div(totalBorrowShares);
+
+    const supplyBalance = common.toBigUnit(collateral, collateralToken.decimals);
+    const borrowBalance = common.toBigUnit(loan.toString(), loanToken.decimals);
+    const totalSupply = common.toBigUnit(totalSupplyAssets, collateralToken.decimals);
+    const totalBorrow = common.toBigUnit(totalBorrowAssets, loanToken.decimals);
+
+    const loanTokenPrice = common.toBigUnit(_loanTokenPrice, 8);
+    const collateralTokenPrice = common.toBigUnit(
+      _loanTokenPrice.mul(normalizedCollateralLoanPrice).div(BigNumber.from(10).pow(normalizedCollateralLoanDecimals)),
```

Comment:
Where did this formula come from?

Reply:
The formula is from [morpho blue oracles](https://docs.morpho.org/contracts/morpho-blue/reference/oracles/#price). I've also included the link in the code comments above.